python/lsst/daf/persistence/butler.py

```diff
@@ -136,44 +149,106 @@ def __init__(self, root, mapper=None, **mapperArgs):
-        :return: None
-        """
-        self._cfg = cfg
         self.datasetTypeAliasDict = {}
 
-        self.repository = Repository.makeFromCfg(self._cfg['repoCfg'])
-
         # Always use an empty Persistence policy until we can get rid of it
         persistencePolicy = pexPolicy.Policy()
         self.persistence = Persistence.getPersistence(persistencePolicy)
         self.log = pexLog.Log(pexLog.Log.getDefaultLog(), "daf.persistence.butler")
-
+        
+        if inputs is None:
+            inputs = []
+        elif not hasattr(inputs, '__iter__'):
+            inputs = [inputs]
+        
+        if outputs is None:
+            outputs = []
+        elif not hasattr(outputs, '__iter__'):
+            outputs = [outputs]
+        
+        self.inputs = []
+        self.outputs = [] 
+
+        for cfg in outputs:
+            if cfg['mode'] == 'r':
+                raise RuntimeError('Output repositories must not be read only.')
+            # All inputs become parents of each output, so add them to each output cfg:
+            cfg['parentCfgs'].extend(inputs)
+            # Outputs must not be read-only:
+            if cfg['mode'] == 'r':
+                raise RuntimeError('Output repositoires must not be read only.')
+            repoData = RepoData(cfg)
+            # readable outputs are also used as inputs:
+            self.outputs.append(repoData)
+            if cfg['mode'] == 'rw':
+                self._addInputs(repoData.cfg)
+
+        for cfg in inputs:
+            self._addInputs(cfg)
+
+        for repoData in self.outputs + self.inputs:
+            repoData.repo = Repository.makeFromCfg(repoData.cfg)
+
+    def _addInputs(self, cfg, tags=[]):
+        """Add a RepoData for cfg if it is not already represented in self.inputs."""
+        # Inputs must not be write-only:
+        if cfg['mode'] == 'w':
+            raise RuntimeError('Input repositoires must not be write only.')
+        repoData = None
+        tags = set(tags)
+        tags.update(set(cfg['tags']))
+
+        # If there is already an output RepoData instance for a given cfg, don't create a new one. Add it to
+        # the inputs list if it's not there, and update its tags.
+        for d in self.outputs:
+            if d.cfg == cfg:
+                self.inputs.append(d)
+                repoData = d
+
+        # If there is already an input RepoData instance for a given cfg, don't create a new one, just update 
+        # its tags.
+        if not repoData:
+            for d in self.inputs:
+                if d.cfg == cfg:
+                    repoData = d
+                    break
+
+        # If no repoData exists for the cfg yet, then create a new one.
+        if not repoData:
+            repoData = RepoData(cfg, tags)
+            self.inputs.append(repoData)
+        else:
+            repoData.tags.update(tags)
+        # add the parents to the input list
```

Comment:
You might document here that this effectively sets the search order for inputs to be depth-first.  There's also a lot of wasted effort because all of the original input repos are added as parents to each output repo before the readable output repos are passed to this method: each original input repo (and all of its parents) has one addition attempt for each readable output repo plus one for itself as an input repo, only one attempt of which will actually do anything (except possibly for adding tags).